src/gcp_airflow_foundations/operators/airflow/external_task.py

```diff
@@ -0,0 +1,157 @@
+import datetime
+import os
+from typing import Any, Callable, FrozenSet, Iterable, Optional, Union
+
+from sqlalchemy import func
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperatorLink, DagBag, DagModel, DagRun, TaskInstance
+from airflow.operators.dummy import DummyOperator
+from airflow.sensors.base import BaseSensorOperator
+from airflow.utils.helpers import build_airflow_url_with_query
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+
+
+class TableIngestionSensor(BaseSensorOperator):
+    """
+    Waits for table ingestion DAGs to complete for a
+    specific execution_date
+
+    :param external_source_tables: A map whose keys are the sourcs you 
+        want to wait for and the values are a list of the tables for each source
+    :type external_source_tables: dict
+    :param allowed_states: Iterable of allowed states, default is ``['success']``
+    :type allowed_states: Iterable
+    :param failed_states: Iterable of failed or dis-allowed states, default is ``None``
+    :type failed_states: Iterable
+    :param execution_delta: time difference with the previous execution to
+        look at, the default is the same execution_date as the current task or DAG.
+        For yesterday, use [positive!] datetime.timedelta(days=1). Either
+        execution_delta or execution_date_fn can be passed to
+        ExternalTaskSensor, but not both.
+    :type execution_delta: Optional[datetime.timedelta]
+    """
+
+    template_fields = ['external_source_tables']
+    ui_color = '#19647e'
+
+    def __init__(
+        self,
+        *,
+        external_source_tables: dict,
+        allowed_states: Optional[Iterable[str]] = None,
+        failed_states: Optional[Iterable[str]] = None,
+        execution_delta: Optional[datetime.timedelta] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.allowed_states = list(allowed_states) if allowed_states else [State.SUCCESS]
+        self.failed_states = list(failed_states) if failed_states else []
+
+        total_states = self.allowed_states + self.failed_states
+        total_states = set(total_states)
+
+        if set(self.failed_states).intersection(set(self.allowed_states)):
+            raise AirflowException(
+                f"Duplicate values provided as allowed "
+                f"`{self.allowed_states}` and failed states `{self.failed_states}`"
+            )
+       
+        self.execution_delta = execution_delta
+        self.external_dag_ids = self.get_external_dag_ids()
+
+    @provide_session
+    def poke(self, context, session=None):
+        if self.execution_delta:
+            dttm = context['execution_date'] - self.execution_delta
+        else:
+            dttm = context['execution_date']
+
+        dttm_filter = dttm if isinstance(dttm, list) else [dttm]
```

Comment:
when would it be a list?

Reply:
Looking at it again, I don't think it would ever be a list. I've updated the line to reflect that.